### PR TITLE
[qunat][graphmode][fx] Add a separate lower_to_native_backend function for relu

### PR DESCRIPTION
**Pull request by jerryzh168**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #63828
* #63627
* #63501
* __->__ #62861
* #62698

Summary:
This PR adds a lower_to_native_backend function to lower a quantized reference model
to a model that uses fbgemm/qnnpack ops. We'll gradually add support and remove
the fbgemm/qnnpack specific handling in quantization_patterns.py

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30165828](https://our.internmc.facebook.com/intern/diff/D30165828)

### CODE REVIEW

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30165828).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30165828).

File: torch/quantization/fx/lower_to_native_backend.py [hunk 1]
```diff
@@ -0,0 +1,8 @@
+import torch
+
+def lower_to_native_backend(model: torch.fx.GraphModule):
```
Comment (vkuzo) [inline torch/quantization/fx/lower_to_native_backend.py:3]: should the type be `nn.Module`, and we trace again?

should we create `lower_to_fbgemm` and `lower_to_qnnpack`, even if the code is the same for now?  They may diverge in the future, and this way the API is more future proof.

Comment (jerryzh168) [reply torch/quantization/fx/lower_to_native_backend.py:3]: Sure, we can trace again.

Yeah I can split the function into 2.

Comment (jerryzh168) [reply torch/quantization/fx/lower_to_native_backend.py:3]: actually if we trace again we might need to carry over configurations from quantization, which is a little bit cumbersome, can we use this as an internal function and just accept torch.fx.GraphModule instead?

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30165828).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30165828).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30165828).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30165828).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30165828).

File: torch/quantization/fx/convert.py [hunk 2]
```diff
@@ -45,6 +45,8 @@
     activation_dtype,
 )
 
+from .lower_to_fbgemm import lower_to_fbgemm
```
Comment (raghuramank100) [inline torch/quantization/fx/convert.py:48]: The convert function here (L145) is almost 400 lines long. Can you refactor it in a later PR? With multiple functions defined inside, it is hard to understand the flow here.

File: torch/quantization/fx/lower_to_native_backend.py (hunk 1 above)
Comment (raghuramank100) [reply torch/quantization/fx/lower_to_native_backend.py:3]: I think it is ok to have this be an internal function as users interface with this only via convert_fx.

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30165828).

File: torch/quantization/fx/convert.py (hunk 2 above)
Comment (jerryzh168) [reply torch/quantization/fx/convert.py:48]: yes sure, will do

Comment (jerryzh168) [reply torch/quantization/fx/convert.py:48]: I think we also might be able to simplify a bit after we change the flow to produce reference pattern by default

File: torch/quantization/fx/quantization_patterns.py [hunk 3]
```diff
@@ -1496,7 +1496,9 @@ def convert(self,
                 load_arg: Callable,
                 is_reference: bool = False,
                 convert_custom_config_dict: Dict[str, Any] = None) -> Node:
-        if is_reference:
+        # always produce reference pattern for relu
```
Comment (raghuramank100) [inline torch/quantization/fx/quantization_patterns.py:1499]: Why should we do this? If we lower to fbgemm or qnnpack, isnt this incorrect?

File: torch/quantization/fx/lower_to_native_backend.py [hunk 4]
```diff
@@ -0,0 +1,8 @@
...
+    """ Lower a quantized reference model (with reference quantized operator patterns)
+    to the native backend in PyTorch (fbgemm/qnnpack), both backends shares the same
+    operator signature so they can be lowered with the same function
+    """
+    pass
```
Comment (raghuramank100) [inline torch/quantization/fx/lower_to_native_backend.py:8]: Currently these functions are no-ops. So how do we go from reference to actual backend?

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30165828).

File: torch/quantization/fx/lower_to_native_backend.py (hunk 4 above)
Comment (jerryzh168) [reply torch/quantization/fx/lower_to_native_backend.py]: sorry, the changes are lost somehow, I just updated the PR again

File: torch/quantization/fx/quantization_patterns.py (hunk 3 above)
Comment (jerryzh168) [reply torch/quantization/fx/quantization_patterns.py:1499]: oh, this is to change convert_fx always to produce reference pattern, and we'll fuse the pattern in lower_to_fbgemm step

Comment (vkuzo) [reply torch/quantization/fx/quantization_patterns.py:1499]: would this break anything for customers who are not using the new `lower_to_fbgemm` function yet?

Comment (jerryzh168) [reply torch/quantization/fx/quantization_patterns.py:1499]: no, this is internal implementation change, users are not explicitly calling lower_to_fbgemm currently. (they just need to call convert_fx)

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30165828).

File: torch/quantization/fx/quantization_patterns.py (hunk 3 above)
Comment (vkuzo) [reply torch/quantization/fx/quantization_patterns.py:1499]: I'm probably confused, let's say a customer has a relu in their model, is_reference is False, and they are not using any lowering functions.  Would this PR change the behavior of calling convert on their model?  From the code it seems like yes, but I might be missing something.

Comment (jerryzh168) [reply torch/quantization/fx/quantization_patterns.py:1499]: no, currently user do not need to call lowering function separately, it's called in convert_fx when is_reference=False.

File: torch/quantization/fx/convert.py [hunk 5]
```diff
@@ -535,4 +537,5 @@ def load_arg_remove(a: Argument) -> Argument:
     model = QuantizedGraphModule(model, act_post_process_removed_graph, preserved_attributes)
     if not is_reference:
         model = fold_weight(model, node_name_to_scope)
+        model = lower_to_fbgemm(model)
```
Comment (jerryzh168) [inline torch/quantization/fx/convert.py:540]: we call the lowering function here in convert_fx, so no change of api for users with this PR

Comment (vkuzo) [reply torch/quantization/fx/convert.py:540]: makes sense, thanks.  Should we add a TODO here to lower to the correct choice from (fbgemm, qnnpack), and also verify that all qconfigs are valid for the selected backend from (fbgemm, qnnpack)?

File: torch/quantization/fx/quantization_patterns.py (hunk 3 above)
Comment (vkuzo) [reply torch/quantization/fx/quantization_patterns.py:1499]: makes sense, thanks!

File: torch/quantization/fx/quantized_fusion_patterns_and_replacements.py [hunk 6]
```diff
@@ -0,0 +1,31 @@
+import torch
+
+def relu_inplace_pattern(x, scale, zero_point):
```
Comment (vkuzo) [inline torch/quantization/fx/quantized_fusion_patterns_and_replacements.py:3]: would be great to have I/O typehints for these

**Review by vkuzo — APPROVED** [review]

**Pull request closed without merging**